app/questionnaire/router.py

```diff
@@ -1,67 +1,141 @@
+from app.helpers.path_finder_helper import path_finder
+from app.questionnaire.location import Location
+
+
 class Router:
-    def __init__(
-        self, schema, routing_path, current_location=None, completed_locations=None
-    ):
+    def __init__(self, schema, completed_store=None):
         self._schema = schema
-        self._routing_path = routing_path
-        self._current_location = current_location
-        self._completed_locations = completed_locations
+        self._completed_store = completed_store
 
-    def can_access_location(self):
+    def can_access_location(self, location: Location, routing_path):
         """
-        Checks whether the current location is valid and accessible.
+        Checks whether the location is valid and accessible.
         :return: boolean
         """
-        if (
-            self._is_current_location_valid()
-            and not self._get_survey_redirect_location()
-        ):
+        allowable_path = self._get_allowable_path(routing_path)
+        if location in allowable_path:
+            block = self._schema.get_block(location.block_id)
+            if (
+                block['type'] in ['Confirmation', 'Summary']
+                and not self.is_survey_complete()
+            ):
+                return False
+
             return True
 
         return False
 
-    def get_next_location(self):
+    def get_next_location(self, location, routing_path):
         """
         Get the first incomplete block in section/survey if trying to access the section/survey end,
         and the section/survey is incomplete or gets the next default location if the above is false.
-        :return: boolean
         """
-        return self._get_survey_redirect_location() or self._get_default_location()
+        current_block_type = self._schema.get_block(location.block_id)['type']
+        last_block_location = routing_path[-1]
+        last_block_type = self._schema.get_block(last_block_location.block_id)['type']
 
-    def _is_current_location_valid(self):
-        """
-        Checks whether the current location is within the routing path.
-        :return: boolean
-        """
-        return self._current_location in self._routing_path
+        # If the section is complete and contains a SectionSummary, return the SectionSummary location
+        if (
+            last_block_type == 'SectionSummary'
+            and current_block_type != last_block_type
```

Comment:
Sorry if I have misunderstood but `current_block_type == last_block_type` for `SectionSummary` can never happen on a POST unless you had a SectionSummary followed by another SectionSummary which wouldn't be the case?

Reply:
It can because there is a submit button on a SectionSummary. Without this check you get stuck in a loop on a SectionSummary.

Reply:
Ah forgot about that button. 👍